### PR TITLE
[wdspec] Navigation aborted event

### PR DESCRIPTION
**Pull request by sadym-chromium**
Navigation should abort previous navigation. Specification:
* [To navigate...](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigate)
* Step 22.3 Queue a global task on the navigation and traversal task source given navigable's active window to **[abort a document and its descendants](https://html.spec.whatwg.org/#abort-a-document-and-its-descendants)** given navigable's active document.
* [To abort a document and its descendants given a Document document](https://html.spec.whatwg.org/#abort-a-document-and-its-descendants)
* Step 3.1 [Abort](https://html.spec.whatwg.org/#abort-a-document) descendantNavigable's active document.
* [To abort a Document document](https://html.spec.whatwg.org/#abort-a-document)
* Step 3.1 Invoke [WebDriver BiDi navigation aborted](https://w3c.github.io/webdriver-bidi/#webdriver-bidi-navigation-aborted).

### CODE REVIEW

File: webdriver/tests/bidi/browsing_context/navigation_aborted/navigation_aborted.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+NAVIGATION_ABORTED_EVENT = "browsingContext.navigationAborted"
+NAVIGATION_STARTED_EVENT = "browsingContext.navigationStarted"
+USER_PROMPT_OPENED_EVENT = "browsingContext.userPromptOpened"
+
+
+async def test_unsubscribe(bidi_session, inline, new_tab):
+    await bidi_session.session.subscribe(events=[NAVIGATION_ABORTED_EVENT])
+    await bidi_session.session.unsubscribe(events=[NAVIGATION_ABORTED_EVENT])
+
+    # Track all received browsingContext.navigationFailed events in the events array.
```
Comment (OrKoN) [inline webdriver/tests/bidi/browsing_context/navigation_aborted/navigation_aborted.py:21]:
```suggestion
    # Track all received browsingContext.navigationAborted events in the events array.
```

File: webdriver/tests/bidi/browsing_context/navigation_aborted/navigation_aborted.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+):
+    slow_page_url = url(
+        "/webdriver/tests/bidi/browsing_context/support/empty.html?pipe=trickle(d10)"
+    )
+    await subscribe_events(events=[NAVIGATION_ABORTED_EVENT])
+
+    result = await bidi_session.browsing_context.navigate(
+        context=new_tab["context"], url=slow_page_url, wait="none"
+    )
+    on_navigation_failed = wait_for_event(NAVIGATION_ABORTED_EVENT)
```
Comment (OrKoN) [inline webdriver/tests/bidi/browsing_context/navigation_aborted/navigation_aborted.py:63]:
```suggestion
    on_navigation_aborted = wait_for_event(NAVIGATION_ABORTED_EVENT)
```

File: webdriver/tests/bidi/browsing_context/navigation_failed/navigation_failed.py [hunk 3]
```diff
@@ -221,87 +221,6 @@ async def on_event(method, data):
     remove_listener()
 
 
-async def test_with_new_navigation(
```
Comment (OrKoN) [inline webdriver/tests/bidi/browsing_context/navigation_failed/navigation_failed.py:224]: should we keep the tests checking that the failed event is not emitted?

**Comment by OrKoN** [comment]
> Navigation should abort previous navigation. Specification:
> 
> * [To navigate...](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigate)
> * Step 22.3 Queue a global task on the navigation and traversal task source given navigable's active window to **[abort a document and its descendants](https://html.spec.whatwg.org/#abort-a-document-and-its-descendants)** given navigable's active document.
> * [To abort a document and its descendants given a Document document](https://html.spec.whatwg.org/#abort-a-document-and-its-descendants)
> * Step 3.1 [Abort](https://html.spec.whatwg.org/#abort-a-document) descendantNavigable's active document.
> * [To abort a Document document](https://html.spec.whatwg.org/#abort-a-document)
> * Step 3.1 Invoke [WebDriver BiDi navigation aborted](https://w3c.github.io/webdriver-bidi/#webdriver-bidi-navigation-aborted).

I am not sure if it applies to this test. Since the previous navigation is still ongoing there is no document created yet? I would need more time to verify that path.

**Comment by OrKoN** [comment]
Yeah, it appears that abort would only be possible after https://html.spec.whatwg.org/multipage/browsing-the-web.html#populating-a-session-history-entry:loading-a-document is done (after the fetch is done).

**Comment by sadym-chromium** [comment]
closing, as I misunderstood the spec

**Pull request closed without merging**